### PR TITLE
refactor: update retry function

### LINKED ISSUE #6447: Update retry function
Current behavior of `retry` function is confusing, we have a option called `retries` but it actually means attempts and not retries, i.e. `retries=1`, will just call the function once.

https://github.com/ChainSafe/lodestar/blob/4c09412ed2c8a94b585d5dba1fe94bc9e2d56a35/packages/utils/src/retry.ts#L27

We should consider renaming this option to either `attempts` or follow more common pattern used in other libraries such as [retry](https://www.npmjs.com/package/retry) and [retry-request](https://www.npmjs.com/package/retry-request) which if `retries=1` will call the function and on failure will retry it once.

### PR DESCRIPTION
**Pull request by FaybianB**
**Motivation**

Update `retry` function.

**Description**

Updates the `retry` function to follow a more common pattern used in other libraries such as [retry](https://www.npmjs.com/package/retry) and [retry-request](https://www.npmjs.com/package/retry-request), which, if `retries=1` it will call the function and on failure retry it once.

Resolves  #6447

Closes #6447

### CODE REVIEW

**Review by nflaig — COMMENTED** [review]
@FaybianB Thanks for looking into this. The change looks good to me, but it would be good to check all callers of this function and see if we need to reduce retries passed by 1. I would expect it doesn't matter much for most operation but getting an overview would be great.

**Comment by FaybianB** [comment]
@nflaig Here are all the functions I've found that call the `retry` function. A lot of the supplied retry values are `20` and `60`. Do you think any of these should be subtracted by 1?:

- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/api/src/utils/client/httpClient.ts#L187-L192
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/cli/test/utils/validator.ts#L64
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/cli/test/e2e/runDevCmd.test.ts#L32
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/cli/test/e2e/voluntaryExit.test.ts#L47-L54
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/cli/test/e2e/voluntaryExit.test.ts#L80-L92
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/utils/test/unit/retry.test.ts#L43
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/cli/test/e2e/voluntaryExitRemoteSigner.test.ts#L67-L74
- https://github.com/ChainSafe/lodestar/blob/efb684a918fffe0efcc04b9a77d5c05caa9cbc6c/packages/cli/test/e2e/voluntaryExitRemoteSigner.test.ts#L95-L107
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/cli/test/e2e/voluntaryExitFromApi.test.ts#L54-L61
- https://github.com/ChainSafe/lodestar/blob/efb684a918fffe0efcc04b9a77d5c05caa9cbc6c/packages/cli/test/e2e/voluntaryExitFromApi.test.ts#L81-L93
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/cli/test/e2e/blsToExecutionchange.test.ts#L45-L52
- https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/spec-test-util/src/downloadTests.ts#L69-L93

**Comment by nflaig** [comment]
Thanks @FaybianB for looking into it

We should be fine to keep the 20/60 as is, but might wanna update these

https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/spec-test-util/src/downloadTests.ts#L88

https://github.com/ChainSafe/lodestar/blob/0e9fc495a442da9759ebe21e82d49d4c00eb0e2d/packages/beacon-node/src/execution/builder/http.ts#L121

I think 3 requests in total should be sufficient here, and we keep previous behavior

**Comment by nflaig** [comment]
~~Just noticed we use external dependency in download tests instead of our own retry implementation. Good opportunity to clean up this dependency~~
https://github.com/ChainSafe/lodestar/blob/c7c99412c3bbe9a550811db7b812f5e6fc741357/packages/spec-test-util/src/downloadTests.ts#L9

https://github.com/ChainSafe/lodestar/blob/533db97f18c22d8ec410820447cecf44c65fdacd/packages/spec-test-util/package.json#L66

**Edit:** Noticed this requires more changes, I will push a PR for this separately (see https://github.com/ChainSafe/lodestar/pull/6458)

**Comment by nflaig** [comment]
@FaybianB I noticed few more cases, we use the retry function in our http clients which are configured via `retryAttempts`, need to double check these as well

e.g. default value here should probably be changed to 2
https://github.com/ChainSafe/lodestar/blob/6b01690f47b358a4d589d7b62933b0b8e7dd8245/packages/beacon-node/src/execution/engine/http.ts#L75

and  we need to make sure to set default value here to 0 (meaning no retries by default)
https://github.com/ChainSafe/lodestar/blob/6b01690f47b358a4d589d7b62933b0b8e7dd8245/packages/beacon-node/src/eth1/provider/jsonRpcHttpClient.ts#L169

(or maybe more elegant would be to not wrap call into `retry` function if `opts?.retryAttempts ?? this.opts?.retryAttempts` is not set / undefined

**Comment by FaybianB** [comment]
@nflaig I've updated the other areas referring to `retryAttempts`. Also, I've added an additional check in the `retry` function to check if we're on the last attempt and bypasses the `shouldRetry` logic if so:

https://github.com/ChainSafe/lodestar/pull/6451/files#diff-7425121b21e5d20801cc486ce948c1ceb550d27819e3fd95177b894c20114e4cR39-R43

I thought this was a cleaner solution than updating the `retry` tests to subtract `1` from the `retryCount`:

https://github.com/ChainSafe/lodestar/blob/d1dc217d0018f926cc92eabc32908fbdefce6cb9/packages/beacon-node/test/e2e/eth1/jsonRpcHttpClient.test.ts#L202

Do you agree or do you think I should update the tests instead?

File: packages/spec-test-util/src/downloadTests.ts [hunk 1]
```diff
@@ -85,7 +85,7 @@ export async function downloadGenericSpecTests<TestNames extends string>(
           log(`Downloaded  ${url}`);
         },
         {
-          retries: 2,
+          retries: 3,
```
Comment (nflaig) [inline packages/spec-test-util/src/downloadTests.ts]: This used [async-retry](https://www.npmjs.com/package/async-retry) library, keeping retries as is, to not change previous behavior

**Comment by nflaig** [comment]
> Do you agree or do you think I should update the tests instead?

Good catch! This makes sense to me, I noticed there is another case where checking if we are at max attempts is important because otherwise, we will also sleep one more time before throwing the last error which is not ideal.

> I thought this was a cleaner solution than updating the retry tests to subtract 1 from the retryCount:

The tests are incorrect, incrementing retryCount in the last iteration is wrong as it will not do another retry. Also noticed tests use retryCount to track request count, pushed some updates to the test to fix this and make them pass.

**Review by nflaig — APPROVED** [review]
LGTM, thanks @FaybianB for the thoroughness!

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.17.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.17.0) :tada:

**Pull request merged**